### PR TITLE
Write GitHub availability warnings to log and console

### PR DESCRIPTION
**Pull request by synthead**
Closes https://github.com/github/octoshift/issues/7006!

This PR adds a quick availability check using our GitHub availability API (https://www.githubstatus.com/api) to determine if there is an unresolved availability incident.  If so, this message appears in the console:

image

If githubstatus.com does not respond with expected data, the console will the warning below and continue the migration.  This effectively adds exception handling if/when GitHub's availability page has availability issues (insert spiderman meme here), but this ensures that customers can continue their migrations as normal if this happens.

image

Please note that this PR copies some prior art from https://github.com/github/gh-gei/pull/331 for checking that GEI/bbs2gh/ado2gh is on the latest version.  This prior art did not include any unit tests (see the PR for details), so no tests have been added in this PR, either.  If tests are important to those who are reviewing this PR, it'd probably be best to file a tech debt issue to handle all the console output in Program.cs for GEI/bbs2gh/ado2gh :+1: 

- [ ] Did you write/update appropriate tests
- [x] Release notes updated (if appropriate)
- [x] Appropriate logging output
- [x] Issue linked
- [x] Docs updated (or issue created)
- [x] New package licenses are added to `ThirdPartyNotices.txt` (if applicable)

### CODE REVIEW

**Comment by dylan-smith** [comment]
I don't think a Factory is necessary here, you should be able to register the Api/Client classes directly with the DI container. Just to make sure I wasn't missing something I tried it and it worked, you can see my changes here: https://github.com/github/gh-gei/commit/197d1ff5b38265e97c376e2f7ad0527e8134e7f9

in this branch: https://github.com/github/gh-gei/tree/dylan-smith/github-status-suggestions

There's a couple other places where we had to do one (or two) http calls and we didn't create client classes in those cases but put the HttpClient code right in the "API" class. Look at VersionChecker and/or HttpDownloadService. HOWEVER, it's probably better to refactor out that common code to a Client class (setting the user-agent headers, logging request/response properly, etc). Rather than naming it GithubStatusClient I'd suggest naming it something generic like StandardHttpClient, or OctoHttpClient or something, then we could use it from VersionChecker, HttpDownloadService, and possibly other things in the future too.

**Comment by synthead** [comment]
> I don't think a Factory is necessary here, you should be able to register the Api/Client classes directly with the DI container. Just to make sure I wasn't missing something I tried it and it worked, you can see my changes here: [197d1ff](https://github.com/github/gh-gei/commit/197d1ff5b38265e97c376e2f7ad0527e8134e7f9)
> 
> in this branch: https://github.com/github/gh-gei/tree/dylan-smith/github-status-suggestions
> 
> There's a couple other places where we had to do one (or two) http calls and we didn't create client classes in those cases but put the HttpClient code right in the "API" class. Look at VersionChecker and/or HttpDownloadService. HOWEVER, it's probably better to refactor out that common code to a Client class (setting the user-agent headers, logging request/response properly, etc). Rather than naming it GithubStatusClient I'd suggest naming it something generic like StandardHttpClient, or OctoHttpClient or something, then we could also use it from VersionChecker, HttpDownloadService, and possibly other things in the future too.

I'm glad you picked up on this!  I added a factory simply due to convention, but I'm 100% happy to remove it!  I co-authored you in f71c1f30546979163bc747eb49f906f75499ad49 that removes the factory and pushed up the changes! :+1:

File: src/ado2gh/Program.cs [hunk 1]
```diff
@@ -54,6 +55,16 @@
...
 
+            try
+            {
+                await GithubStatusCheck(serviceProvider);
+            }
+            catch (Exception ex)
+            {
+                Logger.LogWarning("Could not check GitHub availability from githubstatus.com.  See https://www.githubstatus.com for details.");
+                Logger.LogVerbose(ex.ToString());
+            }
```
[the remark this answers is not in the document]
Comment (synthead) [reply src/ado2gh/Program.cs:67]: This is copying prior art from here!  Happy to change this if we'd like :+1: 

https://github.com/github/gh-gei/blob/f71c1f30546979163bc747eb49f906f75499ad49/src/ado2gh/Program.cs#L69-L77

File: src/bbs2gh/Program.cs [hunk 2]
```diff
@@ -65,6 +66,16 @@
...
 
+            try
+            {
+                await GithubStatusCheck(serviceProvider);
+            }
+            catch (Exception ex)
+            {
+                Logger.LogWarning("Could not check GitHub availability from githubstatus.com.  See https://www.githubstatus.com for details.");
+                Logger.LogVerbose(ex.ToString());
+            }
```
[the remark this answers is not in the document]
Comment (synthead) [reply src/bbs2gh/Program.cs:78]: This is copying prior art from here! Happy to change this if we'd like :+1: 

https://github.com/github/gh-gei/blob/f71c1f30546979163bc747eb49f906f75499ad49/src/bbs2gh/Program.cs#L80-L88

File: src/gei/Program.cs [hunk 3]
```diff
@@ -63,6 +64,16 @@
...
 
+            try
+            {
+                await GithubStatusCheck(serviceProvider);
+            }
+            catch (Exception ex)
+            {
+                Logger.LogWarning("Could not check GitHub availability from githubstatus.com.  See https://www.githubstatus.com for details.");
+                Logger.LogVerbose(ex.ToString());
+            }
```
[the remark this answers is not in the document]
Comment (synthead) [reply src/gei/Program.cs:76]: This is copying prior art from here, too :smile: 

https://github.com/github/gh-gei/blob/f71c1f30546979163bc747eb49f906f75499ad49/src/gei/Program.cs#L78-L86

**Comment by dylan-smith** [comment]
Should probably add a few tests for `BasicHttpClient`. You should be able to mostly copy similar ones from `GithubClientTests`. Specifically these ones:
```
GetAsync_Returns_String_Response
GetAsync_Logs_The_Url
GetAsync_Logs_The_Response_Status_Code_And_Content
GetAsync_Throws_HttpRequestException_On_Non_Success_Response
It_Sets_User_Agent_Header_With_Comments
It_Only_Sets_The_Product_Name_In_User_Agent_Header_When_Version_Provider_Is_Null
```

**Comment by synthead** [comment]
> Should probably add a few tests for `BasicHttpClient`. You should be able to mostly copy similar ones from `GithubClientTests`. Specifically these ones:
> 
> ```
> GetAsync_Returns_String_Response
> GetAsync_Logs_The_Url
> GetAsync_Logs_The_Response_Status_Code_And_Content
> GetAsync_Throws_HttpRequestException_On_Non_Success_Response
> It_Sets_User_Agent_Header_With_Comments
> It_Only_Sets_The_Product_Name_In_User_Agent_Header_When_Version_Provider_Is_Null
> ```

Thank you for your thoughtful suggestion, @dylan-smith!  We just added some tests in c5ce9ef3dcd851f759a345c87e2a609351d18bc5!  How's this look?

**Review by dylan-smith — APPROVED** [review]

**Pull request merged**